renku/core/workflow/plan_factory.py

```diff
@@ -392,23 +394,48 @@ def add_command_input(
 
         mapped_stream = self.get_stream_mapping_for_value(default_value)
 
-        self.inputs.append(
-            CommandInput(
-                id=CommandInput.generate_id(
-                    plan_id=self.plan_id,
-                    position=position,
-                    postfix=mapped_stream.stream_type if mapped_stream else postfix,
-                ),
-                default_value=default_value,
-                prefix=prefix,
+        inp_param = CommandInput(
+            id=CommandInput.generate_id(
+                plan_id=self.plan_id,
                 position=position,
-                mapped_to=mapped_stream,
-                encoding_format=encoding_format,
-                postfix=postfix,
-                name=name,
-            )
+                postfix=mapped_stream.stream_type if mapped_stream else postfix,
+            ),
+            default_value=default_value,
+            prefix=prefix,
+            position=position,
+            mapped_to=mapped_stream,
+            encoding_format=encoding_format,
+            postfix=postfix,
+            name=name,
         )
 
+        existing_parameter = next((p for p in self.inputs if p.name == inp_param.name), None)
+
+        if existing_parameter is not None and existing_parameter.default_value == inp_param.default_value:
+            # merge with new parameter
+            if existing_parameter.prefix is None:
+                existing_parameter.prefix = inp_param.prefix
+
+            if existing_parameter.position is None:
+                existing_parameter.position = inp_param.position
+
+            if existing_parameter.postfix is None:
+                existing_parameter.postfix = inp_param.postfix
+
+            if existing_parameter.encoding_format is None:
+                existing_parameter.encoding_format = inp_param.encoding_format
+
+            if existing_parameter.mapped_to is None:
+                existing_parameter.mapped_to = inp_param.mapped_to
+        elif existing_parameter is not None:
+            # duplicate with different values!
+            raise errors.ParameterError(
+                f"Duplicate input '{inp_param.name}' found with differing values ('{inp_param.default_value}'"
+                f" vs. '{existing_parameter.default_value}')"
+            )
+        else:
+            self.inputs.append(inp_param)
```

Comment:
This is repeated futher but for an output parameter. I guess if we repeat it one or two more times we should factor it out... Just an observation I dont think we need to do it now.

Reply:
they're slightly different, one also has a `create_folder` property (and CommandParameter doesn't have encoding_format or mapped_to).

But it's a good point anyways, it shouldn't be in this code. I refactored it into methods on the type itself.